From 1964df55a1aa3e296bd654ab3158809e1d8af699 Mon Sep 17 00:00:00 2001
From: "Kevin F. Haggerty" <haggertk@lineageos.org>
Date: Mon, 25 Dec 2017 15:30:21 -0700
Subject: [PATCH 11/16] [DO NOT MERGE] klte-common: sepolicy: Rewrite for O

* WIP
* KILL that sepolicy/old/ before merging
* KILL the dontaudits before merging

Change-Id: I6694567fa1c834b262941b9be362c96cbd16625e
---
 BoardConfigCommon.mk                           |  5 +-
 rootdir/etc/init.qcom.rc                       | 19 ++++++--
 rootdir/etc/ueventd.qcom.rc                    |  5 +-
 sepolicy/bluetooth.te                          |  3 --
 sepolicy/cameraserver.te                       |  7 ---
 sepolicy/common/bluetooth.te                   |  7 +++
 sepolicy/common/cameraserver.te                |  2 +
 sepolicy/common/device.te                      |  2 +
 sepolicy/common/dontaudit.te                   | 25 ++++++++++
 sepolicy/common/file.te                        | 16 +++++++
 sepolicy/common/file_contexts                  | 60 ++++++++++++++++++++++++
 sepolicy/common/fsck.te                        |  2 +
 sepolicy/common/genfs_contexts                 |  1 +
 sepolicy/common/hal_fingerprint_default.te     |  6 +++
 sepolicy/common/hal_wifi_default.te            |  4 ++
 sepolicy/common/hal_wifi_supplicant_default.te |  1 +
 sepolicy/common/init.te                        |  5 ++
 sepolicy/common/kernel.te                      |  4 ++
 sepolicy/common/macloader.te                   | 11 +++++
 sepolicy/common/mediaprovider.te               |  2 +
 sepolicy/common/mediaserver.te                 |  4 ++
 sepolicy/common/mm-qcamerad.te                 |  8 ++++
 sepolicy/common/mpdecision.te                  |  2 +
 sepolicy/common/nfc.te                         |  1 +
 sepolicy/common/priv_app.te                    |  5 ++
 sepolicy/common/property_contexts              |  1 +
 sepolicy/common/rild.te                        |  8 ++++
 sepolicy/common/system_server.te               | 12 +++++
 sepolicy/common/tee.te                         |  1 +
 sepolicy/common/thermal-engine.te              |  3 ++
 sepolicy/common/vold.te                        |  2 +
 sepolicy/device.te                             |  1 -
 sepolicy/file.te                               |  8 ----
 sepolicy/file_contexts                         | 63 --------------------------
 sepolicy/fingerprintd.te                       | 14 ------
 sepolicy/fsck.te                               |  1 -
 sepolicy/genfs_contexts                        |  1 -
 sepolicy/healthd.te                            |  2 -
 sepolicy/hostapd.te                            |  1 -
 sepolicy/init.te                               |  3 --
 sepolicy/kernel.te                             |  2 -
 sepolicy/macloader.te                          | 13 ------
 sepolicy/mediaserver.te                        |  6 ---
 sepolicy/mm-qcamerad.te                        |  6 ---
 sepolicy/mpdecision.te                         |  2 -
 sepolicy/old/bluetooth.te                      |  3 ++
 sepolicy/old/cameraserver.te                   |  6 +++
 sepolicy/old/file_contexts                     |  8 ++++
 sepolicy/old/fingerprintd.te                   | 12 +++++
 sepolicy/old/fsck.te                           |  1 +
 sepolicy/old/healthd.te                        |  2 +
 sepolicy/old/hostapd.te                        |  1 +
 sepolicy/old/init.te                           |  3 ++
 sepolicy/old/kernel.te                         |  2 +
 sepolicy/old/macloader.te                      |  9 ++++
 sepolicy/old/mediaserver.te                    |  6 +++
 sepolicy/old/mm-qcamerad.te                    |  6 +++
 sepolicy/old/mpdecision.te                     |  2 +
 sepolicy/old/platform_app.te                   |  4 ++
 sepolicy/old/priv_app.te                       |  5 ++
 sepolicy/old/rild.te                           |  5 ++
 sepolicy/old/shell.te                          |  4 ++
 sepolicy/old/system_server.te                  | 10 ++++
 sepolicy/old/tee.te                            |  2 +
 sepolicy/old/thermal-engine.te                 |  1 +
 sepolicy/old/ueventd.te                        |  6 +++
 sepolicy/old/untrusted_app.te                  |  5 ++
 sepolicy/old/vold.te                           |  1 +
 sepolicy/platform_app.te                       |  4 --
 sepolicy/priv_app.te                           |  5 --
 sepolicy/private/mediaextractor.te             |  2 +
 sepolicy/property_contexts                     | 10 ----
 sepolicy/rild.te                               |  5 --
 sepolicy/sepolicy.mk                           | 28 ++++++++++++
 sepolicy/shell.te                              |  4 --
 sepolicy/system_server.te                      | 10 ----
 sepolicy/tee.te                                |  2 -
 sepolicy/thermal-engine.te                     |  1 -
 sepolicy/ueventd.te                            |  6 ---
 sepolicy/untrusted_app.te                      |  5 --
 sepolicy/vold.te                               |  1 -
 sepolicy/wpa.te                                |  4 --
 82 files changed, 348 insertions(+), 200 deletions(-)
 delete mode 100644 sepolicy/bluetooth.te
 delete mode 100644 sepolicy/cameraserver.te
 create mode 100644 sepolicy/common/bluetooth.te
 create mode 100644 sepolicy/common/cameraserver.te
 create mode 100644 sepolicy/common/device.te
 create mode 100644 sepolicy/common/dontaudit.te
 create mode 100644 sepolicy/common/file.te
 create mode 100644 sepolicy/common/file_contexts
 create mode 100644 sepolicy/common/fsck.te
 create mode 100644 sepolicy/common/genfs_contexts
 create mode 100644 sepolicy/common/hal_fingerprint_default.te
 create mode 100644 sepolicy/common/hal_wifi_default.te
 create mode 100644 sepolicy/common/hal_wifi_supplicant_default.te
 create mode 100644 sepolicy/common/init.te
 create mode 100644 sepolicy/common/kernel.te
 create mode 100644 sepolicy/common/macloader.te
 create mode 100644 sepolicy/common/mediaprovider.te
 create mode 100644 sepolicy/common/mediaserver.te
 create mode 100644 sepolicy/common/mm-qcamerad.te
 create mode 100644 sepolicy/common/mpdecision.te
 create mode 100644 sepolicy/common/nfc.te
 create mode 100644 sepolicy/common/priv_app.te
 create mode 100644 sepolicy/common/property_contexts
 create mode 100644 sepolicy/common/rild.te
 create mode 100644 sepolicy/common/system_server.te
 create mode 100644 sepolicy/common/tee.te
 create mode 100644 sepolicy/common/thermal-engine.te
 create mode 100644 sepolicy/common/vold.te
 delete mode 100644 sepolicy/device.te
 delete mode 100644 sepolicy/file.te
 delete mode 100644 sepolicy/file_contexts
 delete mode 100644 sepolicy/fingerprintd.te
 delete mode 100644 sepolicy/fsck.te
 delete mode 100644 sepolicy/genfs_contexts
 delete mode 100644 sepolicy/healthd.te
 delete mode 100644 sepolicy/hostapd.te
 delete mode 100644 sepolicy/init.te
 delete mode 100644 sepolicy/kernel.te
 delete mode 100644 sepolicy/macloader.te
 delete mode 100644 sepolicy/mediaserver.te
 delete mode 100644 sepolicy/mm-qcamerad.te
 delete mode 100644 sepolicy/mpdecision.te
 create mode 100644 sepolicy/old/bluetooth.te
 create mode 100644 sepolicy/old/cameraserver.te
 create mode 100644 sepolicy/old/file_contexts
 create mode 100644 sepolicy/old/fingerprintd.te
 create mode 100644 sepolicy/old/fsck.te
 create mode 100644 sepolicy/old/healthd.te
 create mode 100644 sepolicy/old/hostapd.te
 create mode 100644 sepolicy/old/init.te
 create mode 100644 sepolicy/old/kernel.te
 create mode 100644 sepolicy/old/macloader.te
 create mode 100644 sepolicy/old/mediaserver.te
 create mode 100644 sepolicy/old/mm-qcamerad.te
 create mode 100644 sepolicy/old/mpdecision.te
 create mode 100644 sepolicy/old/platform_app.te
 create mode 100644 sepolicy/old/priv_app.te
 create mode 100644 sepolicy/old/rild.te
 create mode 100644 sepolicy/old/shell.te
 create mode 100644 sepolicy/old/system_server.te
 create mode 100644 sepolicy/old/tee.te
 create mode 100644 sepolicy/old/thermal-engine.te
 create mode 100644 sepolicy/old/ueventd.te
 create mode 100644 sepolicy/old/untrusted_app.te
 create mode 100644 sepolicy/old/vold.te
 delete mode 100644 sepolicy/platform_app.te
 delete mode 100644 sepolicy/priv_app.te
 create mode 100644 sepolicy/private/mediaextractor.te
 delete mode 100644 sepolicy/property_contexts
 delete mode 100644 sepolicy/rild.te
 create mode 100644 sepolicy/sepolicy.mk
 delete mode 100644 sepolicy/shell.te
 delete mode 100644 sepolicy/system_server.te
 delete mode 100644 sepolicy/tee.te
 delete mode 100644 sepolicy/thermal-engine.te
 delete mode 100644 sepolicy/ueventd.te
 delete mode 100644 sepolicy/untrusted_app.te
 delete mode 100644 sepolicy/vold.te
 delete mode 100644 sepolicy/wpa.te

diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index 9b00077..9fa6517 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -100,10 +100,7 @@ BOARD_RECOVERY_SWIPE := true
 TARGET_RECOVERY_FSTAB := $(COMMON_PATH)/rootdir/etc/fstab.full
 
 # SELinux
--include device/qcom/sepolicy/sepolicy.mk
-
-BOARD_SEPOLICY_DIRS += \
-    $(COMMON_PATH)/sepolicy
+include $(COMMON_PATH)/sepolicy/sepolicy.mk
 
 # Sensors
 TARGET_NO_SENSOR_PERMISSION_CHECK := true
diff --git a/rootdir/etc/init.qcom.rc b/rootdir/etc/init.qcom.rc
index be7d7e4..7f7c6ed 100644
--- a/rootdir/etc/init.qcom.rc
+++ b/rootdir/etc/init.qcom.rc
@@ -111,6 +111,10 @@ on post-fs-data
     chmod 0660 /efs/wifi/.mac.info
     restorecon /efs/wifi/.mac.info
 
+    # for WIFI Front End Module
+    chown system system /data/.cid.info
+    chmod 0664 /data/.cid.info
+
     # Create directory used by audio subsystem
     mkdir /data/misc/audio 0770 audio audio
 
@@ -283,6 +287,14 @@ on boot
     chown radio system /efs/bluetooth
     chmod 0775 /efs/bluetooth
 
+    # Audience ES705 UART
+    # Note: DO NOT move this to ueventd.qcom.rc. Samsung thoughtfully has
+    #       the kernel write directly to /dev/ttyHS3 (!) to load the audience
+    #       firmware. Setting ownership immediately with ueventd would require
+    #       allowing the kernel dac_override, which is an sepolicy neverallow.
+    chmod 0660 /dev/ttyHS3
+    chown media audio /dev/ttyHS3
+
     #Create QMUX deamon socket area
     mkdir /dev/socket/qmux_radio 0770 radio radio
     chmod 2770 /dev/socket/qmux_radio
@@ -849,6 +861,7 @@ on boot
 
     # Set permissions for firmware path control
     chown wifi wifi /sys/module/dhd/parameters/firmware_path
+    chown wifi wifi /sys/module/dhd/parameters/nvram_path
 
 # Services begin here
 
@@ -926,15 +939,13 @@ service wpa_supplicant /vendor/bin/hw/wpa_supplicant \
 service macloader /vendor/bin/macloader
     class late_start
     oneshot
+    user system
+    group system wifi
     seclabel u:r:macloader:s0
 
 on property:wlan.driver.status=ok
     start macloader
 
-on property:init.svc.macloader=stopped
-    chown system root /data/.cid.info
-    chmod 0664 /data/.cid.info
-
 on property:sys.boot_completed=1
     setprop sys.io.scheduler bfq
 
diff --git a/rootdir/etc/ueventd.qcom.rc b/rootdir/etc/ueventd.qcom.rc
index dbfa9a5..e7d5c30 100644
--- a/rootdir/etc/ueventd.qcom.rc
+++ b/rootdir/etc/ueventd.qcom.rc
@@ -147,8 +147,9 @@
 /dev/i2c-5                0660   media      media
 /dev/voice_svc            0660   system     audio
 
-#Audience ES705 UART
-/dev/ttyHS3               0660   media      audio
+# Audience ES705 UART - do not be tempted to uncomment the below unless you
+#   want firmware loading to be blocked by selinux
+#/dev/ttyHS3               0660   media      audio
 
 # Bluetooth
 /dev/ttyHS0               0660   bluetooth  bluetooth
diff --git a/sepolicy/bluetooth.te b/sepolicy/bluetooth.te
deleted file mode 100644
index 2ea6924..0000000
--- a/sepolicy/bluetooth.te
+++ /dev/null
@@ -1,3 +0,0 @@
-allow bluetooth bluetooth_device:chr_file rw_file_perms;
-allow bluetooth proc_bluetooth_writable:dir search;
-allow bluetooth wifi_data_file:file r_file_perms;
diff --git a/sepolicy/cameraserver.te b/sepolicy/cameraserver.te
deleted file mode 100644
index 2a2100a..0000000
--- a/sepolicy/cameraserver.te
+++ /dev/null
@@ -1,7 +0,0 @@
-allow cameraserver camera_socket:sock_file write;
-allow cameraserver init:unix_stream_socket connectto;
-allow cameraserver property_socket:sock_file write;
-allow cameraserver sysfs_camera:dir search;
-allow cameraserver sysfs_camera:file { open read };
-allow cameraserver system_file:file execmod;
-
diff --git a/sepolicy/common/bluetooth.te b/sepolicy/common/bluetooth.te
new file mode 100644
index 0000000..cae1201
--- /dev/null
+++ b/sepolicy/common/bluetooth.te
@@ -0,0 +1,7 @@
+allow bluetooth bluetooth_device:chr_file rw_file_perms;
+allow bluetooth bt_fw_file:file r_file_perms;
+allow bluetooth firmware_file:dir r_dir_perms;
+allow bluetooth proc_bt_sleep:dir search;
+allow bluetooth proc_bt_sleep:file w_file_perms;
+allow bluetooth sysfs_bt_rfkill_state:file w_file_perms;
+allow bluetooth wifi_data_file:file r_file_perms;
diff --git a/sepolicy/common/cameraserver.te b/sepolicy/common/cameraserver.te
new file mode 100644
index 0000000..e3c1e8f
--- /dev/null
+++ b/sepolicy/common/cameraserver.te
@@ -0,0 +1,2 @@
+allow cameraserver camera_socket:sock_file w_file_perms;
+allow cameraserver vendor_file:file execmod;
diff --git a/sepolicy/common/device.te b/sepolicy/common/device.te
new file mode 100644
index 0000000..eef944e
--- /dev/null
+++ b/sepolicy/common/device.te
@@ -0,0 +1,2 @@
+type bluetooth_device, dev_type;
+type efs_block_device, dev_type;
diff --git a/sepolicy/common/dontaudit.te b/sepolicy/common/dontaudit.te
new file mode 100644
index 0000000..a5d655d
--- /dev/null
+++ b/sepolicy/common/dontaudit.te
@@ -0,0 +1,25 @@
+# These will be deleted before committing, I just don't want to keep
+# seeing them during policy bringup
+
+dontaudit shell kernel:system syslog_read;
+
+#dontaudit system_server dalvikcache_data_file:file execute;
+
+dontaudit untrusted_app net_dns_prop:file { open read };
+dontaudit untrusted_app proc:file r_file_perms;
+
+dontaudit untrusted_app_25 camera_prop:file r_file_perms;
+dontaudit untrusted_app_25 debugfs:file r_file_perms;
+dontaudit untrusted_app_25 hal_memtrack_hwservice:hwservice_manager find;
+dontaudit untrusted_app_25 mnt_media_rw_file:dir r_dir_perms;
+dontaudit untrusted_app_25 proc:file r_file_perms;
+dontaudit untrusted_app_25 proc_stat:file r_file_perms;
+dontaudit untrusted_app_25 rootfs:dir r_file_perms;
+dontaudit untrusted_app_25 selinuxfs:file r_file_perms;
+dontaudit untrusted_app_25 serialno_prop:file r_file_perms;
+dontaudit untrusted_app_25 sysfs:file { r_file_perms setattr };
+dontaudit untrusted_app_25 sysfs_devices_system_cpu:file setattr;
+dontaudit untrusted_app_25 sysfs_lowmemorykiller:dir search;
+dontaudit untrusted_app_25 sysfs_lowmemorykiller:file r_file_perms;
+dontaudit untrusted_app_25 userdata_block_device:blk_file getattr;
+dontaudit untrusted_app_25 usermodehelper:file r_file_perms;
diff --git a/sepolicy/common/file.te b/sepolicy/common/file.te
new file mode 100644
index 0000000..ce827f8
--- /dev/null
+++ b/sepolicy/common/file.te
@@ -0,0 +1,16 @@
+type proc_bt_sleep, fs_type;
+
+type sysfs_bt_rfkill_state, fs_type, sysfs_type;
+type sysfs_sec, fs_type, sysfs_type;
+type sysfs_wifi_fw_path, fs_type, sysfs_type;
+type sysfs_wifi_nv_path, fs_type, sysfs_type;
+
+type bt_fw_file, file_type;
+type nfc_fw_file, file_type;
+type vcs_data_file, file_type, data_file_type;
+type wifi_efs_file, file_type;
+
+#type sensors_efs_file, file_type;
+#type sysfs_camera, fs_type, sysfs_type;
+#type sysfs_display, fs_type, sysfs_type;
+#type sysfs_vibeamp, fs_type, sysfs_type;
diff --git a/sepolicy/common/file_contexts b/sepolicy/common/file_contexts
new file mode 100644
index 0000000..86f7091
--- /dev/null
+++ b/sepolicy/common/file_contexts
@@ -0,0 +1,60 @@
+# block devices
+/dev/block/platform/msm_sdcc\.1/by-name/efs             u:object_r:efs_block_device:s0
+/dev/block/platform/msm_sdcc\.1/by-name/fota            u:object_r:misc_block_device:s0
+
+# data files
+/data/.cid.info                                         u:object_r:wifi_data_file:s0
+/data/.wifiver.info                                     u:object_r:wifi_data_file:s0
+/data/(misc|system)/perfd(/.*)?                         u:object_r:mpctl_data_file:s0
+/data/validity(/.*)?                                    u:object_r:vcs_data_file:s0
+
+# device nodes
+/dev/batch_io                                           u:object_r:sensors_device:s0
+/dev/bcm2079x                                           u:object_r:nfc_device:s0
+/dev/btlock                                             u:object_r:bluetooth_device:s0
+/dev/pn547                                              u:object_r:nfc_device:s0
+/dev/rfkill                                             u:object_r:wlan_device:s0
+/dev/sec-nfc                                            u:object_r:nfc_device:s0
+/dev/ttyHS3                                             u:object_r:audio_device:s0
+/dev/vfsspi                                             u:object_r:vcs_device:s0
+
+# efs files
+/efs/bluetooth(/.*)?                                    u:object_r:bluetooth_efs_file:s0
+/efs/wifi(/.*)?                                         u:object_r:wifi_efs_file:s0
+
+# executeables
+/system/vendor/bin/macloader                            u:object_r:macloader_exec:s0
+
+# firmware
+/system/vendor/firmware/bcm4350(.*).hcd                 u:object_r:bt_fw_file:s0
+/system/vendor/firmware/libpn547_fw.so                  u:object_r:nfc_fw_file:s0
+
+# sockets
+/data/cam_socket(.*)                                    u:object_r:camera_socket:s0
+
+# sysfs
+/sys/devices/battery.[0-9]+/power_supply/battery(/.*)?           u:object_r:sysfs_batteryinfo:s0
+/sys/module/dhd/parameters/firmware_path                         u:object_r:sysfs_wifi_fw_path:s0
+/sys/module/dhd/parameters/nvram_path                            u:object_r:sysfs_wifi_nv_path:s0
+/sys/devices/platform/bcm4354_bluetooth/rfkill/rfkill0/state     u:object_r:sysfs_bt_rfkill_state:s0
+/sys/devices/virtual/sec/sec_key/hall_irq_ctrl                   u:object_r:sysfs_sec:s0
+
+# Camera
+#/sys/devices/virtual/camera(/.*)?                       u:object_r:sysfs_camera:s0
+
+# CMHW
+#/sys/devices/virtual/timed_output/vibrator(/.*)?        u:object_r:sysfs_vibeamp:s0
+
+# Display
+#/sys/devices/virtual/lcd/panel/power_reduce             u:object_r:sysfs_display:s0
+
+# Fingerprint
+#/dev/validity(/.*)?                                     u:object_r:vcs_device:s0
+
+# SEC
+#/sys/devices/virtual/sec/tsp(/.*)?                      u:object_r:sysfs_sec:s0
+
+# Sensors
+#/efs/FactoryApp/baro_delta                              u:object_r:sensors_efs_file:s0
+#/efs/gyro_cal_data                                      u:object_r:sensors_efs_file:s0
+#/efs/prox_cal                                           u:object_r:sensors_efs_file:s0
diff --git a/sepolicy/common/fsck.te b/sepolicy/common/fsck.te
new file mode 100644
index 0000000..7f7dcd7
--- /dev/null
+++ b/sepolicy/common/fsck.te
@@ -0,0 +1,2 @@
+allow fsck cache_block_device:blk_file rw_file_perms;
+allow fsck efs_block_device:blk_file rw_file_perms;
diff --git a/sepolicy/common/genfs_contexts b/sepolicy/common/genfs_contexts
new file mode 100644
index 0000000..f74675b
--- /dev/null
+++ b/sepolicy/common/genfs_contexts
@@ -0,0 +1 @@
+genfscon proc /bluetooth/sleep u:object_r:proc_bt_sleep:s0
diff --git a/sepolicy/common/hal_fingerprint_default.te b/sepolicy/common/hal_fingerprint_default.te
new file mode 100644
index 0000000..ae671aa
--- /dev/null
+++ b/sepolicy/common/hal_fingerprint_default.te
@@ -0,0 +1,6 @@
+r_dir_file(hal_fingerprint_default, firmware_file)
+
+allow hal_fingerprint_default tee_device:chr_file rw_file_perms;
+allow hal_fingerprint_default vcs_data_file:dir search;
+allow hal_fingerprint_default vcs_data_file:file rw_file_perms;
+allow hal_fingerprint_default vcs_device:chr_file rw_file_perms;
diff --git a/sepolicy/common/hal_wifi_default.te b/sepolicy/common/hal_wifi_default.te
new file mode 100644
index 0000000..717cf60
--- /dev/null
+++ b/sepolicy/common/hal_wifi_default.te
@@ -0,0 +1,4 @@
+r_dir_file(hal_wifi_default, wifi_efs_file)
+
+allow hal_wifi_default sysfs_wifi_fw_path:file w_file_perms;
+allow hal_wifi_default wifi_data_file:file r_file_perms;
diff --git a/sepolicy/common/hal_wifi_supplicant_default.te b/sepolicy/common/hal_wifi_supplicant_default.te
new file mode 100644
index 0000000..893ea1b
--- /dev/null
+++ b/sepolicy/common/hal_wifi_supplicant_default.te
@@ -0,0 +1 @@
+allow hal_wifi_supplicant_default wlan_device:chr_file r_file_perms;
diff --git a/sepolicy/common/init.te b/sepolicy/common/init.te
new file mode 100644
index 0000000..feb4a25
--- /dev/null
+++ b/sepolicy/common/init.te
@@ -0,0 +1,5 @@
+# Required to load shim libraries
+allow init cameraserver:process { noatsecure };
+allow init hal_camera:process { noatsecure };
+allow init mm-qcamerad:process { noatsecure };
+allow init rild:process { noatsecure };
diff --git a/sepolicy/common/kernel.te b/sepolicy/common/kernel.te
new file mode 100644
index 0000000..8f8e1f1
--- /dev/null
+++ b/sepolicy/common/kernel.te
@@ -0,0 +1,4 @@
+# Samsung literally vfs_write()s to the es705 UART at /dev/ttyHS3 to
+# load the firmware. Without crafting a userspace helper or re-doing
+# the whole path, there is no way around this.
+allow kernel audio_device:chr_file rw_file_perms;
diff --git a/sepolicy/common/macloader.te b/sepolicy/common/macloader.te
new file mode 100644
index 0000000..3642f00
--- /dev/null
+++ b/sepolicy/common/macloader.te
@@ -0,0 +1,11 @@
+type macloader, domain;
+type macloader_exec, exec_type, file_type;
+init_daemon_domain(macloader)
+
+type_transition macloader system_data_file:file wifi_data_file;
+
+r_dir_file(macloader, wifi_efs_file)
+
+allow macloader efs_file:dir search;
+allow macloader sysfs_wifi_nv_path:file w_file_perms;
+allow macloader wifi_data_file:file create_file_perms;
diff --git a/sepolicy/common/mediaprovider.te b/sepolicy/common/mediaprovider.te
new file mode 100644
index 0000000..65ce0b8
--- /dev/null
+++ b/sepolicy/common/mediaprovider.te
@@ -0,0 +1,2 @@
+allow mediaprovider cache_private_backup_file:dir getattr;
+allow mediaprovider cache_recovery_file:dir r_dir_perms;
diff --git a/sepolicy/common/mediaserver.te b/sepolicy/common/mediaserver.te
new file mode 100644
index 0000000..a14d0b3
--- /dev/null
+++ b/sepolicy/common/mediaserver.te
@@ -0,0 +1,4 @@
+allow mediaserver camera_socket:sock_file write;
+allow mediaserver mm-qcamerad:unix_dgram_socket sendto;
+allow mediaserver thermal-engine:unix_stream_socket connectto;
+allow mediaserver vendor_file:file execmod;
diff --git a/sepolicy/common/mm-qcamerad.te b/sepolicy/common/mm-qcamerad.te
new file mode 100644
index 0000000..a004dc5
--- /dev/null
+++ b/sepolicy/common/mm-qcamerad.te
@@ -0,0 +1,8 @@
+type_transition mm-qcamerad system_data_file:sock_file camera_socket;
+
+allow mm-qcamerad camera_socket:sock_file create_file_perms;
+
+# Allow mm-qcamera-daemon to create the socket camera_socket
+allow mm-qcamerad system_data_file:dir w_dir_perms;
+
+allow mm-qcamerad vendor_file:file execmod;
diff --git a/sepolicy/common/mpdecision.te b/sepolicy/common/mpdecision.te
new file mode 100644
index 0000000..94d3f08
--- /dev/null
+++ b/sepolicy/common/mpdecision.te
@@ -0,0 +1,2 @@
+allow mpdecision mpctl_data_file:dir w_dir_perms;
+allow mpdecision mpctl_data_file:sock_file create_file_perms;
diff --git a/sepolicy/common/nfc.te b/sepolicy/common/nfc.te
new file mode 100644
index 0000000..477e977
--- /dev/null
+++ b/sepolicy/common/nfc.te
@@ -0,0 +1 @@
+allow nfc nfc_fw_file:file rx_file_perms;
diff --git a/sepolicy/common/priv_app.te b/sepolicy/common/priv_app.te
new file mode 100644
index 0000000..fe2dc8b
--- /dev/null
+++ b/sepolicy/common/priv_app.te
@@ -0,0 +1,5 @@
+get_prop(priv_app, camera_prop)
+get_prop(priv_app, qemu_hw_mainkeys_prop)
+
+allow priv_app device:dir r_dir_perms;
+allow priv_app proc_interrupts:file r_file_perms;
diff --git a/sepolicy/common/property_contexts b/sepolicy/common/property_contexts
new file mode 100644
index 0000000..05f3ea1
--- /dev/null
+++ b/sepolicy/common/property_contexts
@@ -0,0 +1 @@
+service.camera.hdmi_preview                      u:object_r:camera_prop:s0
diff --git a/sepolicy/common/rild.te b/sepolicy/common/rild.te
new file mode 100644
index 0000000..6bbe2cf
--- /dev/null
+++ b/sepolicy/common/rild.te
@@ -0,0 +1,8 @@
+set_prop(rild, net_radio_prop)
+
+allow rild radio_data_file:dir rw_dir_perms;
+allow rild radio_data_file:file create_file_perms;
+allow rild radio_data_file:lnk_file read;
+
+allow rild proc_net:file w_file_perms;
+allow rild sysfs_sec:file rw_file_perms;
diff --git a/sepolicy/common/system_server.te b/sepolicy/common/system_server.te
new file mode 100644
index 0000000..ca2a9a4
--- /dev/null
+++ b/sepolicy/common/system_server.te
@@ -0,0 +1,12 @@
+get_prop(system_server, alarm_boot_prop)
+
+allow system_server efs_file:dir search;
+allow system_server efs_file:file r_file_perms;
+allow system_server mpctl_data_file:dir search;
+allow system_server mpctl_data_file:sock_file w_file_perms;
+allow system_server mpdecision:unix_stream_socket connectto;
+allow system_server qmuxd:unix_stream_socket connectto;
+allow system_server qmuxd_socket:dir w_dir_perms;
+allow system_server qmuxd_socket:sock_file { create setattr write };
+allow system_server qti_debugfs:file r_file_perms;
+allow system_server sensors_device:chr_file r_file_perms;
diff --git a/sepolicy/common/tee.te b/sepolicy/common/tee.te
new file mode 100644
index 0000000..edb0ac7
--- /dev/null
+++ b/sepolicy/common/tee.te
@@ -0,0 +1 @@
+r_dir_file(tee, vcs_data_file)
diff --git a/sepolicy/common/thermal-engine.te b/sepolicy/common/thermal-engine.te
new file mode 100644
index 0000000..a68d2b0
--- /dev/null
+++ b/sepolicy/common/thermal-engine.te
@@ -0,0 +1,3 @@
+type_transition thermal-engine socket_device:sock_file thermal_socket "thermal-send-client";
+type_transition thermal-engine socket_device:sock_file thermal_socket "thermal-recv-client";
+type_transition thermal-engine socket_device:sock_file thermal_socket "thermal-recv-passive-client";
diff --git a/sepolicy/common/vold.te b/sepolicy/common/vold.te
new file mode 100644
index 0000000..5ce680c
--- /dev/null
+++ b/sepolicy/common/vold.te
@@ -0,0 +1,2 @@
+allow vold efs_file:dir rw_dir_perms;
+allow vold efs_file:file create;
diff --git a/sepolicy/device.te b/sepolicy/device.te
deleted file mode 100644
index 5cc35eb..0000000
--- a/sepolicy/device.te
+++ /dev/null
@@ -1 +0,0 @@
-type bluetooth_device, dev_type;
diff --git a/sepolicy/file.te b/sepolicy/file.te
deleted file mode 100644
index 5ba86a2..0000000
--- a/sepolicy/file.te
+++ /dev/null
@@ -1,8 +0,0 @@
-type sensors_efs_file, file_type;
-type sysfs_camera, fs_type, sysfs_type;
-type sysfs_display, fs_type, sysfs_type;
-type sysfs_sec, fs_type, sysfs_type;
-type sysfs_vibeamp, fs_type, sysfs_type;
-type sysfs_wifi_nv_path, fs_type, sysfs_type;
-type vcs_data_file, file_type, data_file_type;
-type wifi_efs_file, file_type;
diff --git a/sepolicy/file_contexts b/sepolicy/file_contexts
deleted file mode 100644
index 3aaf30c..0000000
--- a/sepolicy/file_contexts
+++ /dev/null
@@ -1,63 +0,0 @@
-# Audience
-/dev/ttyHS3                                             u:object_r:audio_device:s0
-
-# Bluetooth
-/dev/btlock                                             u:object_r:bluetooth_device:s0
-/dev/rfkill                                             u:object_r:bluetooth_device:s0
-/efs/bluetooth(/.*)?                                    u:object_r:bluetooth_efs_file:s0
-
-# Camera
-/data/cam_socket.*                                      u:object_r:camera_socket:s0
-/sys/devices/virtual/camera(/.*)?                       u:object_r:sysfs_camera:s0
-
-# CMHW
-/sys/devices/virtual/timed_output/vibrator(/.*)?        u:object_r:sysfs_vibeamp:s0
-
-# Display
-/sys/devices/virtual/lcd/panel/power_reduce             u:object_r:sysfs_display:s0
-
-# EFS
-/dev/block/platform/msm_sdcc.1/by-name/efs              u:object_r:modem_efs_partition_device:s0
-
-# Fingerprint
-/data/validity(/.*)?                                    u:object_r:vcs_data_file:s0
-/dev/validity(/.*)?                                     u:object_r:vcs_device:s0
-/dev/vfsspi                                             u:object_r:vcs_device:s0
-
-# Macloader
-/system/bin/macloader                                   u:object_r:macloader_exec:s0
-
-# NFC
-/dev/bcm2079x                                           u:object_r:nfc_device:s0
-/dev/pn547                                              u:object_r:nfc_device:s0
-/dev/sec-nfc                                            u:object_r:nfc_device:s0
-
-# RIL
-/data/data/com.android.providers.telephony/databases(/.*)?     u:object_r:radio_data_file:s0
-/data/data/com.android.providers.telephony/shared_prefs(/.*)?  u:object_r:radio_data_file:s0
-
-# RIL - Variant Blobs
-/system/blobs/(.*)/bin/ks                               u:object_r:mdm_helper_exec:s0
-/system/blobs/(.*)/bin/qmuxd                            u:object_r:qmuxd_exec:s0
-/system/blobs/(.*)/bin/rfs_access                       u:object_r:rfs_access_exec:s0
-/system/blobs/(.*)/bin/rild                             u:object_r:rild_exec:s0
-/system/blobs/(.*)/bin/rmt_storage                      u:object_r:rmt_storage_exec:s0
-
-# SEC
-/sys/devices/virtual/sec/sec_key/hall_irq_ctrl          u:object_r:sysfs_sec:s0
-/sys/devices/virtual/sec/tsp(/.*)?                      u:object_r:sysfs_sec:s0
-
-# Sensors
-/dev/batch_io                                           u:object_r:sensors_device:s0
-/efs/FactoryApp/baro_delta                              u:object_r:sensors_efs_file:s0
-/efs/gyro_cal_data                                      u:object_r:sensors_efs_file:s0
-/efs/prox_cal                                           u:object_r:sensors_efs_file:s0
-
-# Uncrypt
-/dev/block/platform/msm_sdcc.1/by-name/fota             u:object_r:misc_block_device:s0
-
-# WiFi
-/data/.cid.info                                         u:object_r:wifi_data_file:s0
-/data/.wifiver.info                                     u:object_r:wifi_data_file:s0
-/efs/wifi(/.*)?                                         u:object_r:wifi_efs_file:s0
-/sys/module/dhd/parameters/nvram_path                   u:object_r:sysfs_wifi_nv_path:s0
diff --git a/sepolicy/fingerprintd.te b/sepolicy/fingerprintd.te
deleted file mode 100644
index 4a25a4c..0000000
--- a/sepolicy/fingerprintd.te
+++ /dev/null
@@ -1,14 +0,0 @@
-allow fingerprintd vcs_data_file:dir create_dir_perms;
-allow fingerprintd vcs_data_file:file create_file_perms;
-
-allow fingerprintd vcs_device:dir create_dir_perms;
-allow fingerprintd vcs_device:file create_file_perms;
-allow fingerprintd vcs_device:fifo_file create_file_perms;
-allow fingerprintd vcs_device:chr_file create_file_perms;
-
-allow fingerprintd tee_device:chr_file rw_file_perms;
-
-allow fingerprintd firmware_file:dir r_dir_perms;
-allow fingerprintd firmware_file:file r_file_perms;
-
-allow fingerprintd vfat:file { getattr open read };
diff --git a/sepolicy/fsck.te b/sepolicy/fsck.te
deleted file mode 100644
index a66c9df..0000000
--- a/sepolicy/fsck.te
+++ /dev/null
@@ -1 +0,0 @@
-allow fsck modem_efs_partition_device:blk_file rw_file_perms;
diff --git a/sepolicy/genfs_contexts b/sepolicy/genfs_contexts
deleted file mode 100644
index 49a35af..0000000
--- a/sepolicy/genfs_contexts
+++ /dev/null
@@ -1 +0,0 @@
-genfscon proc /bluetooth/sleep u:object_r:proc_bluetooth_writable:s0
diff --git a/sepolicy/healthd.te b/sepolicy/healthd.te
deleted file mode 100644
index cdf9e70..0000000
--- a/sepolicy/healthd.te
+++ /dev/null
@@ -1,2 +0,0 @@
-allow healthd device:dir r_dir_perms;
-allow healthd rtc_device:chr_file rw_file_perms;
diff --git a/sepolicy/hostapd.te b/sepolicy/hostapd.te
deleted file mode 100644
index f9388f6..0000000
--- a/sepolicy/hostapd.te
+++ /dev/null
@@ -1 +0,0 @@
-allow hostapd bluetooth_device:chr_file { open read };
diff --git a/sepolicy/init.te b/sepolicy/init.te
deleted file mode 100644
index 9254911..0000000
--- a/sepolicy/init.te
+++ /dev/null
@@ -1,3 +0,0 @@
-allow init sysfs_sec:lnk_file r_file_perms;
-allow init debugfs:file write;
-allow init socket_device:sock_file { create write setattr };
diff --git a/sepolicy/kernel.te b/sepolicy/kernel.te
deleted file mode 100644
index 2193b45..0000000
--- a/sepolicy/kernel.te
+++ /dev/null
@@ -1,2 +0,0 @@
-allow kernel audio_device:chr_file rw_file_perms;
-allow kernel efs_file:dir search;
diff --git a/sepolicy/macloader.te b/sepolicy/macloader.te
deleted file mode 100644
index df90419..0000000
--- a/sepolicy/macloader.te
+++ /dev/null
@@ -1,13 +0,0 @@
-type macloader, domain;
-type macloader_exec, exec_type, file_type;
-init_daemon_domain(macloader)
-
-type_transition macloader system_data_file:file wifi_data_file;
-
-allow macloader efs_file:dir search;
-allow macloader self:capability { chown dac_override fowner fsetid };
-allow macloader sysfs_wifi_nv_path:file { open write };
-allow macloader system_data_file:dir { add_name search write };
-allow macloader wifi_data_file:file { create_file_perms getattr setattr };
-allow macloader wifi_efs_file:dir search;
-allow macloader wifi_efs_file:file r_file_perms;
diff --git a/sepolicy/mediaserver.te b/sepolicy/mediaserver.te
deleted file mode 100644
index 7ebc14c..0000000
--- a/sepolicy/mediaserver.te
+++ /dev/null
@@ -1,6 +0,0 @@
-allow mediaserver cameraproxy_service:service_manager find;
-allow mediaserver sensorservice_service:service_manager find;
-allow mediaserver sysfs_camera:dir search;
-allow mediaserver sysfs_camera:file { getattr open read };
-allow mediaserver system_file:file execmod; # for libmmjpeg
-allow mediaserver system_prop:property_service set;
diff --git a/sepolicy/mm-qcamerad.te b/sepolicy/mm-qcamerad.te
deleted file mode 100644
index f30345c..0000000
--- a/sepolicy/mm-qcamerad.te
+++ /dev/null
@@ -1,6 +0,0 @@
-allow mm-qcamerad media_rw_data_file:dir search;
-allow mm-qcamerad sysfs_camera:dir search;
-allow mm-qcamerad sysfs_camera:file rw_file_perms;
-allow mm-qcamerad system_data_file:dir w_dir_perms;
-allow mm-qcamerad system_file:file execmod; # for libmmcamera_faceproc
-type_transition mm-qcamerad system_data_file:sock_file camera_socket "cam_socket3";
diff --git a/sepolicy/mpdecision.te b/sepolicy/mpdecision.te
deleted file mode 100644
index 25e559f..0000000
--- a/sepolicy/mpdecision.te
+++ /dev/null
@@ -1,2 +0,0 @@
-allow mpdecision system_data_file:dir { add_name remove_name write };
-allow mpdecision system_data_file:sock_file write;
diff --git a/sepolicy/old/bluetooth.te b/sepolicy/old/bluetooth.te
new file mode 100644
index 0000000..2ea6924
--- /dev/null
+++ b/sepolicy/old/bluetooth.te
@@ -0,0 +1,3 @@
+allow bluetooth bluetooth_device:chr_file rw_file_perms;
+allow bluetooth proc_bluetooth_writable:dir search;
+allow bluetooth wifi_data_file:file r_file_perms;
diff --git a/sepolicy/old/cameraserver.te b/sepolicy/old/cameraserver.te
new file mode 100644
index 0000000..11cf7d7
--- /dev/null
+++ b/sepolicy/old/cameraserver.te
@@ -0,0 +1,6 @@
+allow cameraserver camera_socket:sock_file write;
+allow cameraserver init:unix_stream_socket connectto;
+allow cameraserver property_socket:sock_file write;
+allow cameraserver sysfs_camera:dir search;
+allow cameraserver sysfs_camera:file { open read };
+allow cameraserver system_file:file execmod;
diff --git a/sepolicy/old/file_contexts b/sepolicy/old/file_contexts
new file mode 100644
index 0000000..b946283
--- /dev/null
+++ b/sepolicy/old/file_contexts
@@ -0,0 +1,8 @@
+# Camera
+/data/cam_socket.*                                      u:object_r:camera_socket:s0
+
+# EFS
+/dev/block/platform/msm_sdcc.1/by-name/efs              u:object_r:modem_efs_partition_device:s0
+
+# Macloader
+/system/bin/macloader                                   u:object_r:macloader_exec:s0
diff --git a/sepolicy/old/fingerprintd.te b/sepolicy/old/fingerprintd.te
new file mode 100644
index 0000000..0010127
--- /dev/null
+++ b/sepolicy/old/fingerprintd.te
@@ -0,0 +1,12 @@
+allow fingerprintd firmware_file:dir r_dir_perms;
+allow fingerprintd firmware_file:file r_file_perms;
+allow fingerprintd vcs_data_file:dir create_dir_perms;
+allow fingerprintd vcs_data_file:file create_file_perms;
+
+allow fingerprintd tee_device:chr_file rw_file_perms;
+allow fingerprintd vcs_device:dir create_dir_perms;
+allow fingerprintd vcs_device:file create_file_perms;
+allow fingerprintd vcs_device:fifo_file create_file_perms;
+
+allow fingerprintd vcs_device:chr_file create_file_perms;
+allow fingerprintd vfat:file { getattr open read };
diff --git a/sepolicy/old/fsck.te b/sepolicy/old/fsck.te
new file mode 100644
index 0000000..a66c9df
--- /dev/null
+++ b/sepolicy/old/fsck.te
@@ -0,0 +1 @@
+allow fsck modem_efs_partition_device:blk_file rw_file_perms;
diff --git a/sepolicy/old/healthd.te b/sepolicy/old/healthd.te
new file mode 100644
index 0000000..cdf9e70
--- /dev/null
+++ b/sepolicy/old/healthd.te
@@ -0,0 +1,2 @@
+allow healthd device:dir r_dir_perms;
+allow healthd rtc_device:chr_file rw_file_perms;
diff --git a/sepolicy/old/hostapd.te b/sepolicy/old/hostapd.te
new file mode 100644
index 0000000..f9388f6
--- /dev/null
+++ b/sepolicy/old/hostapd.te
@@ -0,0 +1 @@
+allow hostapd bluetooth_device:chr_file { open read };
diff --git a/sepolicy/old/init.te b/sepolicy/old/init.te
new file mode 100644
index 0000000..9254911
--- /dev/null
+++ b/sepolicy/old/init.te
@@ -0,0 +1,3 @@
+allow init sysfs_sec:lnk_file r_file_perms;
+allow init debugfs:file write;
+allow init socket_device:sock_file { create write setattr };
diff --git a/sepolicy/old/kernel.te b/sepolicy/old/kernel.te
new file mode 100644
index 0000000..2193b45
--- /dev/null
+++ b/sepolicy/old/kernel.te
@@ -0,0 +1,2 @@
+allow kernel audio_device:chr_file rw_file_perms;
+allow kernel efs_file:dir search;
diff --git a/sepolicy/old/macloader.te b/sepolicy/old/macloader.te
new file mode 100644
index 0000000..0b3b410
--- /dev/null
+++ b/sepolicy/old/macloader.te
@@ -0,0 +1,9 @@
+type_transition macloader system_data_file:file wifi_data_file;
+
+allow macloader efs_file:dir search;
+allow macloader self:capability { chown dac_override fowner fsetid };
+allow macloader sysfs_wifi_nv_path:file { open write };
+allow macloader system_data_file:dir { add_name search write };
+allow macloader wifi_data_file:file { create_file_perms getattr setattr };
+allow macloader wifi_efs_file:dir search;
+allow macloader wifi_efs_file:file r_file_perms;
diff --git a/sepolicy/old/mediaserver.te b/sepolicy/old/mediaserver.te
new file mode 100644
index 0000000..7ebc14c
--- /dev/null
+++ b/sepolicy/old/mediaserver.te
@@ -0,0 +1,6 @@
+allow mediaserver cameraproxy_service:service_manager find;
+allow mediaserver sensorservice_service:service_manager find;
+allow mediaserver sysfs_camera:dir search;
+allow mediaserver sysfs_camera:file { getattr open read };
+allow mediaserver system_file:file execmod; # for libmmjpeg
+allow mediaserver system_prop:property_service set;
diff --git a/sepolicy/old/mm-qcamerad.te b/sepolicy/old/mm-qcamerad.te
new file mode 100644
index 0000000..f30345c
--- /dev/null
+++ b/sepolicy/old/mm-qcamerad.te
@@ -0,0 +1,6 @@
+allow mm-qcamerad media_rw_data_file:dir search;
+allow mm-qcamerad sysfs_camera:dir search;
+allow mm-qcamerad sysfs_camera:file rw_file_perms;
+allow mm-qcamerad system_data_file:dir w_dir_perms;
+allow mm-qcamerad system_file:file execmod; # for libmmcamera_faceproc
+type_transition mm-qcamerad system_data_file:sock_file camera_socket "cam_socket3";
diff --git a/sepolicy/old/mpdecision.te b/sepolicy/old/mpdecision.te
new file mode 100644
index 0000000..25e559f
--- /dev/null
+++ b/sepolicy/old/mpdecision.te
@@ -0,0 +1,2 @@
+allow mpdecision system_data_file:dir { add_name remove_name write };
+allow mpdecision system_data_file:sock_file write;
diff --git a/sepolicy/old/platform_app.te b/sepolicy/old/platform_app.te
new file mode 100644
index 0000000..42592cd
--- /dev/null
+++ b/sepolicy/old/platform_app.te
@@ -0,0 +1,4 @@
+allow platform_app exfat:dir create_dir_perms;
+allow platform_app exfat:file create_file_perms;
+allow platform_app fuseblk:dir create_dir_perms;
+allow platform_app fuseblk:file create_file_perms;
diff --git a/sepolicy/old/priv_app.te b/sepolicy/old/priv_app.te
new file mode 100644
index 0000000..6b968bd
--- /dev/null
+++ b/sepolicy/old/priv_app.te
@@ -0,0 +1,5 @@
+allow priv_app device:dir { open read };
+allow priv_app exfat:dir create_dir_perms;
+allow priv_app exfat:file create_file_perms;
+allow priv_app fuseblk:dir create_dir_perms;
+allow priv_app fuseblk:file create_file_perms;
diff --git a/sepolicy/old/rild.te b/sepolicy/old/rild.te
new file mode 100644
index 0000000..9be05bc
--- /dev/null
+++ b/sepolicy/old/rild.te
@@ -0,0 +1,5 @@
+allow rild proc_net:file rw_file_perms;
+allow rild self:capability dac_override;
+allow rild sysfs_sec:file rw_file_perms;
+allow rild radio_data_file:lnk_file read;
+allow rild radio_prop:property_service set;
diff --git a/sepolicy/old/shell.te b/sepolicy/old/shell.te
new file mode 100644
index 0000000..4e8e4b6
--- /dev/null
+++ b/sepolicy/old/shell.te
@@ -0,0 +1,4 @@
+allow shell exfat:dir create_dir_perms;
+allow shell exfat:file create_file_perms;
+allow shell fuseblk:dir create_dir_perms;
+allow shell fuseblk:file create_file_perms;
diff --git a/sepolicy/old/system_server.te b/sepolicy/old/system_server.te
new file mode 100644
index 0000000..75c1bd4
--- /dev/null
+++ b/sepolicy/old/system_server.te
@@ -0,0 +1,10 @@
+allow system_server efs_file:dir search;
+allow system_server sensors_efs_file:file r_file_perms;
+allow system_server sysfs_display:file rw_file_perms;
+allow system_server sysfs_sec:dir search;
+allow system_server sysfs_sec:file rw_file_perms;
+allow system_server sysfs_vibeamp:dir search;
+allow system_server sysfs_vibeamp:file rw_file_perms;
+allow system_server wifi_efs_file:dir search;
+allow system_server wifi_efs_file:file r_file_perms;
+allow system_server app_data_file:file rename;
diff --git a/sepolicy/old/tee.te b/sepolicy/old/tee.te
new file mode 100644
index 0000000..02eeef6
--- /dev/null
+++ b/sepolicy/old/tee.te
@@ -0,0 +1,2 @@
+allow tee vcs_data_file:dir create_dir_perms;
+allow tee vcs_data_file:file create_file_perms;
diff --git a/sepolicy/old/thermal-engine.te b/sepolicy/old/thermal-engine.te
new file mode 100644
index 0000000..97ae015
--- /dev/null
+++ b/sepolicy/old/thermal-engine.te
@@ -0,0 +1 @@
+allow thermal-engine self:capability chown;
diff --git a/sepolicy/old/ueventd.te b/sepolicy/old/ueventd.te
new file mode 100644
index 0000000..d03550c
--- /dev/null
+++ b/sepolicy/old/ueventd.te
@@ -0,0 +1,6 @@
+allow ueventd sysfs_camera:file rw_file_perms;
+allow ueventd sysfs_sec:file rw_file_perms;
+allow ueventd sysfs_vibeamp:file rw_file_perms;
+allow ueventd vcs_device:chr_file create_file_perms;
+allow ueventd vfat:dir search;
+allow ueventd vfat:file { getattr open read };
diff --git a/sepolicy/old/untrusted_app.te b/sepolicy/old/untrusted_app.te
new file mode 100644
index 0000000..be13689
--- /dev/null
+++ b/sepolicy/old/untrusted_app.te
@@ -0,0 +1,5 @@
+# These are safe for an untrusted_app -- they are the external SD card
+allow untrusted_app exfat:dir create_dir_perms;
+allow untrusted_app exfat:file create_file_perms;
+allow untrusted_app fuseblk:dir create_dir_perms;
+allow untrusted_app fuseblk:file create_file_perms;
diff --git a/sepolicy/old/vold.te b/sepolicy/old/vold.te
new file mode 100644
index 0000000..61ec258
--- /dev/null
+++ b/sepolicy/old/vold.te
@@ -0,0 +1 @@
+allow vold efs_file:dir r_file_perms;
diff --git a/sepolicy/platform_app.te b/sepolicy/platform_app.te
deleted file mode 100644
index 42592cd..0000000
--- a/sepolicy/platform_app.te
+++ /dev/null
@@ -1,4 +0,0 @@
-allow platform_app exfat:dir create_dir_perms;
-allow platform_app exfat:file create_file_perms;
-allow platform_app fuseblk:dir create_dir_perms;
-allow platform_app fuseblk:file create_file_perms;
diff --git a/sepolicy/priv_app.te b/sepolicy/priv_app.te
deleted file mode 100644
index 6b968bd..0000000
--- a/sepolicy/priv_app.te
+++ /dev/null
@@ -1,5 +0,0 @@
-allow priv_app device:dir { open read };
-allow priv_app exfat:dir create_dir_perms;
-allow priv_app exfat:file create_file_perms;
-allow priv_app fuseblk:dir create_dir_perms;
-allow priv_app fuseblk:file create_file_perms;
diff --git a/sepolicy/private/mediaextractor.te b/sepolicy/private/mediaextractor.te
new file mode 100644
index 0000000..88b7c6d
--- /dev/null
+++ b/sepolicy/private/mediaextractor.te
@@ -0,0 +1,2 @@
+allow mediaextractor exfat:file r_file_perms;
+allow mediaextractor sdcardfs:file r_file_perms;
diff --git a/sepolicy/property_contexts b/sepolicy/property_contexts
deleted file mode 100644
index 1f84c35..0000000
--- a/sepolicy/property_contexts
+++ /dev/null
@@ -1,10 +0,0 @@
-##########################
-# property service keys
-#
-#
-persist.ril.radiocapa.tdscdma            u:object_r:radio_prop:s0
-persist.ril.modem.board                  u:object_r:radio_prop:s0
-persist.ril.ims.eutranParam              u:object_r:radio_prop:s0
-persist.ril.ims.utranParam               u:object_r:radio_prop:s0
-persist.ril.xcap.pdnFailCause            u:object_r:radio_prop:s0
-persist.ril.ims.pdnFailCause             u:object_r:radio_prop:s0
diff --git a/sepolicy/rild.te b/sepolicy/rild.te
deleted file mode 100644
index 9be05bc..0000000
--- a/sepolicy/rild.te
+++ /dev/null
@@ -1,5 +0,0 @@
-allow rild proc_net:file rw_file_perms;
-allow rild self:capability dac_override;
-allow rild sysfs_sec:file rw_file_perms;
-allow rild radio_data_file:lnk_file read;
-allow rild radio_prop:property_service set;
diff --git a/sepolicy/sepolicy.mk b/sepolicy/sepolicy.mk
new file mode 100644
index 0000000..45022e2
--- /dev/null
+++ b/sepolicy/sepolicy.mk
@@ -0,0 +1,28 @@
+#
+# Copyright (C) 2018 The LineageOS Project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+-include device/qcom/sepolicy/sepolicy.mk
+-include device/qcom/sepolicy/legacy-sepolicy.mk
+
+# Board specific SELinux policy variable definitions
+BOARD_SEPOLICY_DIRS += \
+    device/samsung/klte-common/sepolicy/common \
+
+BOARD_PLAT_PUBLIC_SEPOLICY_DIR += \
+    device/samsung/klte-common/sepolicy/public
+
+BOARD_PLAT_PRIVATE_SEPOLICY_DIR += \
+    device/samsung/klte-common/sepolicy/private
diff --git a/sepolicy/shell.te b/sepolicy/shell.te
deleted file mode 100644
index 4e8e4b6..0000000
--- a/sepolicy/shell.te
+++ /dev/null
@@ -1,4 +0,0 @@
-allow shell exfat:dir create_dir_perms;
-allow shell exfat:file create_file_perms;
-allow shell fuseblk:dir create_dir_perms;
-allow shell fuseblk:file create_file_perms;
diff --git a/sepolicy/system_server.te b/sepolicy/system_server.te
deleted file mode 100644
index 75c1bd4..0000000
--- a/sepolicy/system_server.te
+++ /dev/null
@@ -1,10 +0,0 @@
-allow system_server efs_file:dir search;
-allow system_server sensors_efs_file:file r_file_perms;
-allow system_server sysfs_display:file rw_file_perms;
-allow system_server sysfs_sec:dir search;
-allow system_server sysfs_sec:file rw_file_perms;
-allow system_server sysfs_vibeamp:dir search;
-allow system_server sysfs_vibeamp:file rw_file_perms;
-allow system_server wifi_efs_file:dir search;
-allow system_server wifi_efs_file:file r_file_perms;
-allow system_server app_data_file:file rename;
diff --git a/sepolicy/tee.te b/sepolicy/tee.te
deleted file mode 100644
index 02eeef6..0000000
--- a/sepolicy/tee.te
+++ /dev/null
@@ -1,2 +0,0 @@
-allow tee vcs_data_file:dir create_dir_perms;
-allow tee vcs_data_file:file create_file_perms;
diff --git a/sepolicy/thermal-engine.te b/sepolicy/thermal-engine.te
deleted file mode 100644
index 97ae015..0000000
--- a/sepolicy/thermal-engine.te
+++ /dev/null
@@ -1 +0,0 @@
-allow thermal-engine self:capability chown;
diff --git a/sepolicy/ueventd.te b/sepolicy/ueventd.te
deleted file mode 100644
index d03550c..0000000
--- a/sepolicy/ueventd.te
+++ /dev/null
@@ -1,6 +0,0 @@
-allow ueventd sysfs_camera:file rw_file_perms;
-allow ueventd sysfs_sec:file rw_file_perms;
-allow ueventd sysfs_vibeamp:file rw_file_perms;
-allow ueventd vcs_device:chr_file create_file_perms;
-allow ueventd vfat:dir search;
-allow ueventd vfat:file { getattr open read };
diff --git a/sepolicy/untrusted_app.te b/sepolicy/untrusted_app.te
deleted file mode 100644
index be13689..0000000
--- a/sepolicy/untrusted_app.te
+++ /dev/null
@@ -1,5 +0,0 @@
-# These are safe for an untrusted_app -- they are the external SD card
-allow untrusted_app exfat:dir create_dir_perms;
-allow untrusted_app exfat:file create_file_perms;
-allow untrusted_app fuseblk:dir create_dir_perms;
-allow untrusted_app fuseblk:file create_file_perms;
diff --git a/sepolicy/vold.te b/sepolicy/vold.te
deleted file mode 100644
index 61ec258..0000000
--- a/sepolicy/vold.te
+++ /dev/null
@@ -1 +0,0 @@
-allow vold efs_file:dir r_file_perms;
diff --git a/sepolicy/wpa.te b/sepolicy/wpa.te
deleted file mode 100644
index 1cdf25b..0000000
--- a/sepolicy/wpa.te
+++ /dev/null
@@ -1,4 +0,0 @@
-allow wpa bluetooth_device:chr_file rw_file_perms;
-allow wpa efs_file:dir search;
-allow wpa wifi_efs_file:dir search;
-allow wpa wifi_efs_file:file r_file_perms;
-- 
2.7.4

